packages/salesforcedx-vscode-replay-debugger/src/breakpoints/checkpointService.ts

```diff
@@ -774,15 +853,119 @@ export async function sfdxCreateCheckpoints() {
   }
 }
 
+// A couple of important notes about this command's processing
+// 1. There is no way to invoke a breakpoint change through vscode.debug
+//    there is only add/delete.
+// 2. A changed breakpoint has to first be deleted and then re-added.
+// 3. Add/Delete breakpoints will go through the processBreakpointChangedForCheckpoints
+//    event that's registered. That'll take care of all the checkpoint specific processing.
+// 4. When a breakpoint already exists and it is being converted to a checkpoint, only
+//    the hitCondition (which is really the hit count) is kept. The other pieces of information
+//    that may be on the checkpoint are the condition (which needs to get set to Checkpoint)
+//    and the logMessage. The logMessage is scrapped since this ends up being taken over by
+//    checkpoints for user input SOQL or Apex.
+export async function sfdxToggleCheckpoint() {
+  const bpAdd: vscode.Breakpoint[] = [];
+  const bpRemove: vscode.Breakpoint[] = [];
+  const uri = exports.fetchActiveEditorUri();
+  const lineNumber = exports.fetchActiveSelectionLineNumber();
+
+  if (uri && lineNumber) {
+    // While selection could be passed directly into the location instead of creating
+    // a new range, it ends up creating a weird secondary icon on the line with the
+    // breakpoint which is due to the start/end characters being non-zero.
+    let hitCondition = undefined;
+    const bp = fetchExistingBreakpointForUriAndLineNumber(uri, lineNumber);
+    // There's already a breakpoint at this line
+    if (bp) {
+      // If the breakpoint is a checkpoint then remove it and return
+      if (
+        bp.condition &&
+        bp.condition!.toLowerCase().indexOf(CHECKPOINT) >= 0
+      ) {
+        bpRemove.push(bp);
+        return await vscode.debug.removeBreakpoints(bpRemove);
+      } else {
+        // The only thing from the old breakpoint that is applicable to keep is the hitCondition
+        // which maps to iterations. Squirrel away hitCondition, remove the breakpoint and let
+        // processing go into the code to create a new breakpoint with the checkpoint condition
+        hitCondition = bp.hitCondition;
+        bpRemove.push(bp);
+        await vscode.debug.removeBreakpoints(bpRemove);
+      }
+    }
+
+    // Create a new checkpoint/breakpoint from scratch.
+    const range = new vscode.Range(lineNumber, 0, lineNumber, 0);
+    const location = new vscode.Location(uri, range);
+    const newBreakpoint = new vscode.SourceBreakpoint(
+      location,
+      true,
+      CHECKPOINT,
+      hitCondition
+    );
+    bpAdd.push(newBreakpoint);
+    await vscode.debug.addBreakpoints(bpAdd);
+  }
+  return;
+}
+
+// This methods was broken out of sfdxToggleCheckpoint for testing purposes.
+function fetchActiveEditorUri(): vscode.Uri | undefined {
+  const editor = vscode.window.activeTextEditor;
+  if (editor) {
+    return editor.document.uri;
+  }
+}
+exports.fetchActiveEditorUri = fetchActiveEditorUri;
```

Comment:
Normally we use the `export` keyword with the function if it's something that needs to be shared across classes so line 932 would be something `function fetchActiveEditorUri(): vscode.Uri | undefined {`

This should let you call them directly without having to do `export.function`

Reply:
As per our conversation this is the way it is in order to be able to sinon.stub the method for testing.